src/kvstore.c

```diff
@@ -165,16 +165,27 @@ static dict *createDictIfNeeded(kvstore *kvs, int didx) {
     return kvs->dicts[didx];
 }
 
-static void freeDictIfNeeded(kvstore *kvs, int didx) {
-    if (!(kvs->flags & KVSTORE_FREE_EMPTY_DICTS) ||
-        !kvstoreGetDict(kvs, didx) ||
+/* Free the dict if the dict is empty. Unlike freeDictIfNeeded which is handled
+ * implicitly by kvstore and depends on the configuration, this one is meant to
+ * be an explicit decision by the caller who may know the dict will no longer
+ * be needed. */
+void kvstoreFreeDictIfEmpty(kvstore *kvs, int didx) {
+    if (!kvstoreGetDict(kvs, didx) ||
         kvstoreDictSize(kvs, didx) != 0)
         return;
+
     dictRelease(kvs->dicts[didx]);
     kvs->dicts[didx] = NULL;
     kvs->allocated_dicts--;
 }
 
+/* If KVSTORE_FREE_EMPTY_DICTS is set and the dict is empty, free the dict. */
+static void freeDictIfNeeded(kvstore *kvs, int didx) {
```

Comment:
```suggestion
void freeDictIfNeeded(kvstore *kvs, int didx) {
```
we should avoid this method not being inlined, instead, let `kvstoreFreeDictIfEmpty()` be inlined as much as possible.
maybe it's better give it `__attribute__ ((noinline))`.

before this PR
```
  throughput summary: 21840.74 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
       73.069     1.616    72.959    76.735    79.743    91.519
```

this PR
```
  throughput summary: 21779.85 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
       73.277     1.632    73.343    76.799    78.975    90.623
```

this PR (remove static)
```
  throughput summary: 22620.34 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
       70.558     1.616    70.463    73.407    76.799    87.743
```

Reply:
@sundb what did you benchmark? is it @enjoy-binbin's last script (with he loop for 100 incr and del)?

from what i remember static or no static isn't suppose to affect inlining.
what am i missing?

in any case these minor differences in benchamrk result are not interesting.
maybe someone can modify redis-benchmark and run one with a good pipeline and no Lua,
it's all very odd to me that these extra mallocs don't make a bigger difference.

Reply:
> @sundb what did you benchmark? is it @enjoy-binbin's last script (with he loop for 100 incr and del)?
> 

same script except `-n 500000`
```./src/redis-benchmark -n 500000 -P 32 eval "for i=1,100 do redis.call('incr', KEYS[1]) redis.call('del', KEYS[1]) end" 1 key```

> from what i remember static or no static isn't suppose to affect inlining. what am i missing?

`static` does affect the compiler and is more likely to be inlined, i confirm this with `disas`.